File: module-app/src/main/java/maple/expectation/scheduler/PopularCharacterWarmupScheduler.java
package maple.expectation.scheduler;

import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Timer;
import java.time.LocalDateTime;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import maple.expectation.core.port.out.CacheWarmupPort;
import maple.expectation.core.port.out.PopularCharacterTrackerPort;
import maple.expectation.infrastructure.executor.LogicExecutor;
import maple.expectation.infrastructure.executor.TaskContext;
import maple.expectation.infrastructure.lock.LockStrategy;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

/**
 * 인기 캐릭터 자동 웜업 스케줄러 (#275 Auto Warmup)
 *
 * <h3>기능</h3>
 *
 * <p>전날 인기 캐릭터 TOP N을 조회하여 캐시를 미리 웜업합니다. 서버 시작 시 또는 매일 새벽에 실행되어 Cold Start 문제를 해결합니다.
 *
 * <h3>실행 시점</h3>
 *
 * <ul>
 *   <li>매일 새벽 5시 (트래픽 낮은 시간)
 *   <li>서버 시작 후 30초 (초기 웜업)
 * </ul>
 *
 * <h3>5-Agent Council 합의</h3>
 *
 * <ul>
 *   <li>Green (Performance): 웜업으로 첫 요청 지연 방지, RPS 향상
 *   <li>Blue (Architect): 분산 락으로 단일 인스턴스만 웜업 실행
 *   <li>Red (SRE): 요청 간 지연으로 Thundering Herd 방지
 *   <li>Yellow (QA): 성공/실패 메트릭으로 모니터링
 * </ul>
 *
 * <h3>설정</h3>
 *
 * <pre>
 * scheduler:
 *   warmup:
 *     enabled: true           # 웜업 활성화 여부
 *     top-count: 50           # 웜업할 상위 캐릭터 수
 *     delay-between-ms: 100   # 요청 간 지연 (ms)
 * </pre>
 *
 * @see PopularCharacterTracker 인기 캐릭터 트래커
 */
@Slf4j
@Component
@ConditionalOnProperty(
    name = "scheduler.warmup.enabled",
    havingValue = "true",
    matchIfMissing = false // 명시적으로 활성화 필요
    )
@RequiredArgsConstructor
public class PopularCharacterWarmupScheduler {

  private final PopularCharacterTrackerPort popularCharacterTracker;
  private final CacheWarmupPort cacheWarmupPort;
  private final LockStrategy lockStrategy;
  private final LogicExecutor executor;
  private final MeterRegistry meterRegistry;

  @Value("${scheduler.warmup.top-count:50}")
  private int topCount;

  @Value("${scheduler.warmup.delay-between-ms:100}")
  private long delayBetweenMs;

  // Stable gauge instances (registered once, updated each run)
  private final AtomicInteger successCount = new AtomicInteger(0);
  private final AtomicInteger failCount = new AtomicInteger(0);

  /**
   * Constructor with gauge registration (once at startup)
   *
   * <p>Lombok's @RequiredArgsConstructor generates the constructor, so we use @PostConstruct for
   * gauge initialization instead of a custom constructor.
   */
  @jakarta.annotation.PostConstruct
  public void init() {
    // Register gauges once to stable instances
    meterRegistry.gauge("warmup.last.success_count", successCount);
    meterRegistry.gauge("warmup.last.fail_count", failCount);
  }

  /**
   * 매일 새벽 5시 웜업 실행
   *
   * <p>트래픽이 낮은 시간에 캐시를 미리 채워둡니다.
   */
  @Scheduled(cron = "0 0 5 * * *")
  public void dailyWarmup() {
    executeWarmup("DailyWarmup");
  }

  /**
   * 서버 시작 후 30초 뒤 초기 웜업
   *
   * <p>서버 재시작 후 Cold Cache 상태를 빠르게 해소합니다.
   */
  @Scheduled(initialDelay = 30000, fixedDelay = Long.MAX_VALUE)
  public void initialWarmup() {
    executeWarmup("InitialWarmup");
  }

  /**
   * 웜업 실행 (분산 락 사용)
   *
   * <p>여러 인스턴스 중 하나만 웜업을 실행하도록 분산 락을 사용합니다.
   *
   * @param warmupType 웜업 유형 (로깅용)
   */
  private void executeWarmup(String warmupType) {
    TaskContext context = TaskContext.of("Scheduler", "Warmup." + warmupType);

    executor.executeOrCatch(
        () -> {
          // 분산 락 획득 시도 (waitTime=0: 즉시 반환, leaseTime=300초)
          lockStrategy.executeWithLock(
              "popular-warmup-lock",
              0,
              300,
              () -> {
                doWarmup(warmupType);
                return null;
              });
          return null;
        },
        e -> {
          if (e instanceof maple.expectation.error.exception.DistributedLockException) {
            log.debug("[Warmup] {} skipped: another instance is warming up", warmupType);
          } else {
            log.error("[Warmup] {} failed: {}", warmupType, e.getMessage());
            meterRegistry
                .counter("warmup.execution", "type", warmupType, "status", "error")
                .increment();
          }
          return null;
        },
        context);
  }

  /**
   * 실제 웜업 로직
   *
   * @param warmupType 웜업 유형
   */
  private void doWarmup(String warmupType) {
    Timer.Sample sample = Timer.start(meterRegistry);
    log.info(
        "[Warmup] {} started at {} - warming up top {} characters",
        warmupType,
        LocalDateTime.now(),
        topCount);

    // 전날 인기 캐릭터 조회
    List<String> topCharacters = popularCharacterTracker.getYesterdayTopCharacters(topCount);

    if (topCharacters.isEmpty()) {
      log.info("[Warmup] {} completed - no characters to warm up (first day?)", warmupType);
      meterRegistry.counter("warmup.execution", "type", warmupType, "status", "empty").increment();
      return;
    }

    // Reset stable counters for new warmup run
    successCount.set(0);
    failCount.set(0);

    for (String userIgn : topCharacters) {
      warmupCharacter(userIgn);

      // Thundering Herd 방지를 위한 요청 간 지연
      if (delayBetweenMs > 0) {
        sleep(delayBetweenMs);
      }
    }

    sample.stop(meterRegistry.timer("warmup.duration", "type", warmupType));
    meterRegistry.counter("warmup.execution", "type", warmupType, "status", "success").increment();

    log.info(
        "[Warmup] {} completed - success: {}, fail: {}, total: {}",
        warmupType,
        successCount.get(),
        failCount.get(),
        topCharacters.size());
  }

  /**
   * 단일 캐릭터 웜업
   *
   * @param userIgn 캐릭터 닉네임
   */
  private void warmupCharacter(String userIgn) {
    executor.executeOrCatch(
        () -> {
          // V4 API 호출하여 캐시 채우기 (force=false로 기존 캐시 사용)
          cacheWarmupPort.warmup(userIgn, false);
          successCount.incrementAndGet();
          log.debug("[Warmup] Warmed up: {}", maskIgn(userIgn));
          return null;
        },
        e -> {
          failCount.incrementAndGet();
          log.warn("[Warmup] Failed to warm up {}: {}", maskIgn(userIgn), e.getMessage());
          return null;
        },
        TaskContext.of("Warmup", "Character", userIgn));
  }

  /**
   * 지연 (Thread.sleep 래핑)
   *
   * <p><b>Section 14 Exception:</b> Thread.sleep is acceptable here for sequential warmup delays
   * with proper LogicExecutor wrapping. This is a synchronous delay in a sequential processing
   * loop, not an asynchronous scheduled task.
   *
   * @param millis 대기 시간 (밀리초)
   */
  private void sleep(long millis) {
    executor.executeOrDefault(
        () -> {
          Thread.sleep(millis);
          return null;
        },
        null,
        TaskContext.of("Warmup", "Delay"));
  }

  /** IGN 마스킹 (로깅용) */
  private String maskIgn(String ign) {
    if (ign == null || ign.length() < 2) return "***";
    return ign.charAt(0) + "***" + ign.substring(ign.length() - 1);
  }
}

File: module-infra/src/main/kotlin/maple/expectation/infrastructure/scheduler/PopularCharacterWarmupScheduler.kt
package maple.expectation.infrastructure.scheduler

import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.Timer
import jakarta.annotation.PostConstruct
import java.time.LocalDateTime
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.LockSupport
import maple.expectation.core.port.out.CacheWarmupPort
import maple.expectation.core.port.out.PopularCharacterTrackerPort
import maple.expectation.error.exception.DistributedLockException
import maple.expectation.infrastructure.executor.LogicExecutor
import maple.expectation.infrastructure.executor.TaskContext
import maple.expectation.infrastructure.lock.LockStrategy
import maple.expectation.util.StringMaskingUtils.maskIgn
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component

/**
 * 인기 캐릭터 자동 웜업 스케줄러 (ADR-005 이관)
 */
@Component
@ConditionalOnProperty(
    name = ["scheduler.warmup.enabled"],
    havingValue = "true",
    matchIfMissing = false,
)
class PopularCharacterWarmupScheduler(
    private val popularCharacterTracker: PopularCharacterTrackerPort,
    private val cacheWarmupPort: CacheWarmupPort,
    private val lockStrategy: LockStrategy,
    private val executor: LogicExecutor,
    private val meterRegistry: MeterRegistry,
) {
    private val log = LoggerFactory.getLogger(PopularCharacterWarmupScheduler::class.java)

    @Value("\${scheduler.warmup.top-count:50}")
    private var topCount: Int = 50

    @Value("\${scheduler.warmup.delay-between-ms:100}")
    private var delayBetweenMs: Long = 100

    private val successCount = AtomicInteger(0)
    private val failCount = AtomicInteger(0)

    @PostConstruct
    fun init() {
        meterRegistry.gauge("warmup.last.success_count", successCount)
        meterRegistry.gauge("warmup.last.fail_count", failCount)
    }

    @Scheduled(cron = "0 0 5 * * *")
    fun dailyWarmup() {
        executeWarmup("DailyWarmup")
    }

    @Scheduled(initialDelay = 30000, fixedDelay = Long.MAX_VALUE)
    fun initialWarmup() {
        executeWarmup("InitialWarmup")
    }

    private fun executeWarmup(warmupType: String) {
        val context = TaskContext.of("Scheduler", "Warmup.$warmupType")

        executor.executeOrCatch(
            {
                lockStrategy.executeWithLock(
                    "popular-warmup-lock",
                    0,
                    300,
                ) {
                    doWarmup(warmupType)
                    null
                }
                null
            },
            { e ->
                if (e is DistributedLockException) {
                    log.debug("[Warmup] {} skipped: another instance is warming up", warmupType)
                } else {
                    log.error("[Warmup] {} failed: {}", warmupType, e.message)
                    meterRegistry.counter("warmup.execution", "type", warmupType, "status", "error")
                        .increment()
                }
                null
            },
            context,
        )
    }

    private fun doWarmup(warmupType: String) {
        val sample = Timer.start(meterRegistry)
        log.info(
            "[Warmup] {} started at {} - warming up top {} characters",
            warmupType,
            LocalDateTime.now(),
            topCount,
        )

        val topCharacters = popularCharacterTracker.getYesterdayTopCharacters(topCount)

        if (topCharacters.isEmpty()) {
            log.info("[Warmup] {} completed - no characters to warm up (first day?)", warmupType)
            meterRegistry.counter("warmup.execution", "type", warmupType, "status", "empty")
                .increment()
            return
        }

        successCount.set(0)
        failCount.set(0)

        for (userIgn in topCharacters) {
            warmupCharacter(userIgn)
            if (delayBetweenMs > 0) {
                sleep(delayBetweenMs)
            }
        }

        sample.stop(meterRegistry.timer("warmup.duration", "type", warmupType))
        meterRegistry.counter("warmup.execution", "type", warmupType, "status", "success")
            .increment()

        log.info(
            "[Warmup] {} completed - success: {}, fail: {}, total: {}",
            warmupType,
            successCount.get(),
            failCount.get(),
            topCharacters.size,
        )
    }

    private fun warmupCharacter(userIgn: String) {
        executor.executeOrCatch(
            {
                cacheWarmupPort.warmup(userIgn, false)
                successCount.incrementAndGet()
                log.debug("[Warmup] Warmed up: {}", maskIgn(userIgn))
                null
            },
            { e ->
                failCount.incrementAndGet()
                log.warn("[Warmup] Failed to warm up {}: {}", maskIgn(userIgn), e.message)
                null
            },
            TaskContext.of("Warmup", "Character", userIgn),
        )
    }

    private fun sleep(millis: Long) {
        LockSupport.parkNanos(TimeUnit.MILLISECONDS.toNanos(millis))
    }
}
